modin/engines/base/frame/data.py

```diff
@@ -976,37 +977,80 @@ def _join_index_objects(self, axis, other_index, how, sort):
-                Whether or not to sort the joined index
+        axis : 0 or 1
+            The axis index object to join (0 - rows, 1 - columns).
+        others_index : list(Index)
+            The others_index to join on.
+        how : {'left', 'right', 'inner', 'outer'}
+            The type of join to join to make.
+        sort : boolean
+            Whether or not to sort the joined index
 
         Returns
         -------
-        Index
-            Joined indices.
+        (Index, func)
+            Joined index with make_reindexer func
         """
+        assert isinstance(others_index, list)
 
-        def merge_index(obj1, obj2):
+        # define helper functions
+        def merge(left_index, right_index):
             if axis == 1 and how == "outer" and not sort:
-                return obj1.union(obj2, sort=False)
+                return left_index.union(right_index, sort=False)
             else:
-                return obj1.join(obj2, how=how, sort=sort)
-
-        if isinstance(other_index, list):
-            joined_obj = self.columns if axis else self.index
-            # TODO: revisit for performance
-            for obj in other_index:
-                joined_obj = merge_index(joined_obj, obj)
-            return joined_obj
-        if axis:
-            return merge_index(self.columns, other_index)
+                return left_index.join(right_index, how=how, sort=sort)
+
+        # define condition for joining indexes
+        do_join_index = False
+        for index in others_index[1:]:
+            if not others_index[0].equals(index):
+                do_join_index = True
+                break
+
+        # define condition for joining indexes with getting indexers
+        is_duplicates = any(not index.is_unique for index in others_index) and axis == 0
+        indexers = []
+        if is_duplicates:
+            indexers = [None] * len(others_index)
+
+        # perform joining indexes
+        if do_join_index:
+            if len(others_index) == 2 and is_duplicates:
+                # in case of count of indexes > 2 we should perform joining all indexes
+                # after that get indexers
+                # in the fast path we can obtain joined_index and indexers in one call
+                joined_index, indexers[0], indexers[1] = others_index[0].join(
+                    others_index[1], how=how, sort=sort, return_indexers=True
+                )
+            else:
+                joined_index = others_index[0]
+                # TODO: revisit for performance
+                for index in others_index[1:]:
+                    joined_index = merge(joined_index, index)
+
+                if is_duplicates:
+                    for i, index in enumerate(others_index):
+                        indexers[i] = index.get_indexer_for(joined_index)
         else:
-            return self.index.join(other_index, how=how, sort=sort)
+            joined_index = others_index[0].copy()
```

Comment:
Why `copy`?

Reply:
The index is then passed to the constructor of the data frame in which no memory is copied. That is, the index is shared between two frames (changes in one - will lead to changes in the other).